vllm/entrypoints/openai/serving_chat.py

```diff
@@ -486,6 +604,23 @@ async def chat_completion_stream_generator(
                                           index=i)
                         ])
 
+                    elif request.tool_choice == "required":
```

Comment:
this is clean, but it would be nice to be able to group the required and auto branch together. 
Wdyt?
Also, is an `extract_tool_calls_streaming(required=True/False)` interface much worse than a separate `extract_tool_call_required_streaming`?

Reply:
As far as I know, the `extract_tool_calls_streaming` calls are methods of specific tool parsers for auto tool call parsing (e.g. Hermes or Llama). I thought about writing a separate tool parser for the required case. A drawback of this that this would either require the vLLM server to be started with a specific tool parser, which is unnecessary, or create a new separate tool parsing method, which is static, in the `ToolParser` super class and use this instead. A `true/false` flag in each method would require implementation in all sub classes that then call the same method, because the format of the required tool call is always the same and determined by guided decoding.

Reply:
makes sense! We could just factor into a `handle_tool_choice` function if code becomes too bloated.